main/src/library/Array.flix

```diff
@@ -14,6 +14,13 @@
  * limitations under the License.
  */
 
+instance Iterable[Array[a, r]] {
+    type Elm = a
+    type Aef = r
+    pub def iterator(rc: Region[r1], a: Array[a, r]): Iterator[a, r + r1, r1] \ (r + r1) =
+        checked_ecast(Array.iterator(rc, a))
```

Comment:
Philosophical musing: It seems that creating an iterator over something shouldn't _actually_ touch that thing; it's only when iterating over it that we touch it. So I would expect this to have only effect `r1` and not `r`.

When I was prototyping associated effects before, that's what I did for the sample collections, though it required some tedious rewriting of the implementations since many of them use `r + r1` for the effect.

Reply:
Ideally `iterator` would probably be "dont touch input in building the iterator" i.e. 

> pub def iterator(rc: Region[r1], a: Array[a, r]): Iterator[a, r + r1, r1] \ r1

But most iterators from mutable data structures are imperative in the "setup" phase as well as the "next/iterate" phase.

Reply:
What works is done in the setup phase?

Reply:
I also know we had some discussion about whether the "middle" argument of the iterator should include the region itself or not.

Reply:
Initializing ref cells that are updated by `next`.